### DISCUSSION TITLE
Welcome to ag4masses Discussions!

### DISCUSSION THREAD

**Discussion by tpgh24** [tpgh24/ag4masses#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

**Comment by bhwangfy** [comment]
Hi, thanks for having this exciting project! The author of the original paper seems to never respond to the issues. I also want to contribute to the project, but it looks like the discussion channel here is barely used. Are there any other ways we can discuss, for example, through WeChat/email or other ways...?

**Reply by tpgh24** [reply]
Thanks for the post. I think Discussion board is the best way to share knowledge and thoughts. Maybe after your post, we will have more people interested in the project to join us. 😊

**Comment by bhwangfy** [comment]
Hi I would like to report a bug -- for the following example
` a b c d = quadrangle a b c d ; e f g h = eqdia_quadrangle e f g h ; i j k l = quadrangle i j k l ; m = parallelogram m l f i ; n o = trisegment n o e g ; p = intersection_cc p h l g ; q r s t = cc_tangent q r s t p d d p ; u = psquare u r p ; v = reflect v a k s ; w = orthocenter w k t m ? eqangle6 s d s t u d u t`
, it can not pass the goal checking in build_problem, but if we hack the build_problem to skip the goal checking and do ddar, it can still  write a proof

**Reply by bhwangfy** [reply]
This is true for the original ag repo, I haven't tried it with the AG4Masses

**Reply by tpgh24** [reply]
Can you post the error message you got?

**Reply by bhwangfy** [reply]
Ah, the goal itself is incorrect (can not pass numerical check) and it is weird that ag can still write a proof for it -- and sometimes running this problem will raise the error 
I0619 10:22:25.920118 140020570691392 graph.py:967] a b c d = quadrangle a b c d; e f g h = eqdia_quadrangle e f g h; i j k l = quadrangle i j k l; m = parallelogram m l f i; n o = trisegment n o e g; p = intersection_cc p h l g; q r s t = cc_tangent q r s t p d d p; u = psquare u r p; v = reflect v a k s; w = orthocenter w k t m ? eqangle6 s d s t u d u t
I0619 10:22:29.068161 140020570691392 ddar.py:61] Depth 1/1000 time = 2.9713425636291504
I0619 10:22:40.101403 140020570691392 ddar.py:61] Depth 2/1000 time = 11.032436609268188
I0619 10:23:06.060093 140020570691392 ddar.py:61] Depth 3/1000 time = 25.958303689956665
Traceback (most recent call last):
  File "alphageometry/alphageometry.py", line 962, in main
    run_ddar(g, this_problem, _OUT_FILE.value)
  File "alphageometry/alphageometry.py", line 320, in run_ddar
    ddar.solve(g, RULES, p, max_level=1000)
  File "alphageometry/ddar.py", line 149, in solve
    dervs, eq4, next_branches, added = saturate_or_goal(
  File "alphageometry/ddar.py", line 51, in saturate_or_goal
    added, derv, eq4, n_branching = dd.bfs_one_level(
  File "alphageometry/dd.py", line 1177, in bfs_one_level
    add = g.add_piece(name, args, deps=deps)
  File "alphageometry/graph.py", line 1204, in add_piece
    return self.add_cyclic(args, deps)
  File "alphageometry/graph.py", line 2074, in add_cyclic
    circle = self.get_new_circle_thru_triplet(p1, p2, p3)
  File "alphageometry/graph.py", line 1375, in get_new_circle_thru_triplet
    circle.num = nm.Circle(p1=p1.num, p2=p2.num, p3=p3.num)
  File "alphageometry/numericals.py", line 407, in __init__
    center = line_line_intersection(l12, l23)
  File "alphageometry/numericals.py", line 583, in line_line_intersection
    raise InvalidLineIntersectError
numericals.InvalidLineIntersectError

**Reply by tpgh24** [reply]
When I run this problem in AG4Masses, I get a different error:

```
I0618 22:54:45.490167 140173838004224 graph.py:499] a b c d = quadrangle a b c d; e f g h = eqdia_quadrangle e f g h; i j k l = quadrangle i j k l; m = parallelogram m l f i; n o = trisegment n o e g; p = intersection_cc p h l g; q r s t = cc_tangent q r s t p d d p; u = psquare u r p; v = reflect v a k s; w = orthocenter w k t m ? cong a b a b
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/peng/ag4masses/alphageometry/alphageometry.py", line 752, in <module>
    app.run(main)
  File "/home/peng/pythovenv/lib/python3.10/site-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/home/peng/pythovenv/lib/python3.10/site-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "/home/peng/ag4masses/alphageometry/alphageometry.py", line 734, in main
    g, _ = gh.Graph.build_problem(this_problem, DEFINITIONS)
  File "/home/peng/ag4masses/alphageometry/graph.py", line 506, in build_problem
    adds, plevel = g.add_clause(
  File "/home/peng/ag4masses/alphageometry/graph.py", line 2567, in add_clause
    args = self.names2points([mapping[a] for a in d.args])
  File "/home/peng/ag4masses/alphageometry/graph.py", line 556, in names2points
    raise ValueError(f'Cannot find point {name} in graph')
ValueError: Cannot find point m in graph
```
This is because the syntax of the construction action `parallelogram` is different from most of other construction actions, its output is the last argument, not the first! So you need to write `m = parallelogram l f i m` instead of `m = parallelogram m l f i`. See the `def.txt` file. I don't know why the authors of AlphaGeometry made it this way, this probably should be considered a bug. There are a few other construction actions in `def.txt` like this.

I think the reason you got the error you posted instead of the one I get is because you ran into another bug in AlphaGeometry that's fixed in AG4Masses, before the hitting the issue I get.

After I fix the `parallelogram` arguments and replace the problem's goal with `cong a b a b`, I can get a diagram, indicating the syntax of the premise part of the problem is correct. But using your original conclusion the code freezes, this usualy indicates that the conclusion is not correct, as you mentioned.
 
Since you hacked the code to skip the problem validity check, ddar may run through the proving process and give an invalid proof to an invalid problem. Do you have the original problem to check whether the conclusion is written correctly?

**Reply by bhwangfy** [reply]
Thanks for the post -- I had manually adjusted the constructor in my code so I didn't run into the "cannot find m" error (sorry I forgot to mention it). Moreover, I just run this example (with goal deleted) using ag4mass, and ddar throws the error

```
 Skipping registering GPU devices...
I0619 16:02:42.198732 140092576634688 graph.py:498] random_problem1
I0619 16:02:42.198983 140092576634688 graph.py:499] 
I0619 16:02:45.220079 140092576634688 ddar.py:60] Depth 1/1000 time = 2.7831528186798096
I0619 16:02:54.909900 140092576634688 ddar.py:60] Depth 2/1000 time = 9.689586162567139
Traceback (most recent call last):
  File "ag4masses/./alphageometry/alphageometry.py", line 735, in main
    run_ddar(g, this_problem, _OUT_FILE.value)
  File "ag4masses/./alphageometry/alphageometry.py", line 227, in run_ddar
    ddar.solve(g, RULES, p, max_level=1000)
  File "ag4masses/alphageometry/ddar.py", line 95, in solve
    dervs, eq4, next_branches, added = saturate_or_goal(
  File "ag4masses/alphageometry/ddar.py", line 50, in saturate_or_goal
    added, derv, eq4, n_branching = dd.bfs_one_level(
  File "ag4masses/alphageometry/dd.py", line 1094, in bfs_one_level
    add = g.add_piece(name, args, deps=deps)
  File "ag4masses/alphageometry/graph.py", line 689, in add_piece
    return self.add_perp(args, deps)
  File "ag4masses/alphageometry/graph.py", line 1268, in add_perp
    raise ValueError(f'{ab.name} and {cd.name} Cannot be perp.')
ValueError: dp and dp Cannot be perp.
```

**Comment by tpgh24** [comment]
I cannot reproduce this error. Are you able to run any problem with AG4Masses?
